src/components/CustomCheckout.tsx

```diff
@@ -0,0 +1,275 @@
+    return null;
+  }
+
+  const {on: _on, session: _session, ...actions} = customCheckoutSdk;
+  return {...actions, ...customCheckoutSdk.session()};
+};
+
+interface CustomCheckoutProviderProps {
+  /**
+   * A [Stripe object](https://stripe.com/docs/js/initializing) or a `Promise` resolving to a `Stripe` object.
+   * The easiest way to initialize a `Stripe` object is with the the [Stripe.js wrapper module](https://github.com/stripe/stripe-js/blob/master/README.md#readme).
+   * Once this prop has been set, it can not be changed.
+   *
+   * You can also pass in `null` or a `Promise` resolving to `null` if you are performing an initial server-side render or when generating a static site.
+   */
+  stripe: PromiseLike<stripeJs.Stripe | null> | stripeJs.Stripe | null;
+  options: stripeJs.StripeCustomCheckoutOptions;
+}
+
+interface PrivateCustomCheckoutProviderProps {
+  stripe: unknown;
+  options: stripeJs.StripeCustomCheckoutOptions;
+  children?: ReactNode;
+}
+const INVALID_STRIPE_ERROR =
+  'Invalid prop `stripe` supplied to `CustomCheckoutProvider`. We recommend using the `loadStripe` utility from `@stripe/stripe-js`. See https://stripe.com/docs/stripe-js/react#elements-props-stripe for details.';
+
+export const CustomCheckoutProvider: FunctionComponent<PropsWithChildren<
+  CustomCheckoutProviderProps
+>> = (({
+  stripe: rawStripeProp,
+  options,
+  children,
+}: PrivateCustomCheckoutProviderProps) => {
+  const parsed = React.useMemo(
+    () => parseStripeProp(rawStripeProp, INVALID_STRIPE_ERROR),
+    [rawStripeProp]
+  );
+
+  // State used to trigger a re-render when sdk.session is updated
+  const [
+    _,
+    setSession,
+  ] = React.useState<stripeJs.StripeCustomCheckoutSession | null>(null);
+
+  // State used to avoid calling initCustomCheckout multiple times when options changes
+  const [
+    initCustomCheckoutCalled,
+    setInitCustomCheckoutCalled,
+  ] = React.useState<boolean>(false);
+
+  const [ctx, setContext] = React.useState<CustomCheckoutSdkContextValue>(
+    () => ({
+      stripe: parsed.tag === 'sync' ? parsed.stripe : null,
+      customCheckoutSdk: null,
+    })
+  );
+
+  const safeSetContext = (
+    stripe: stripeJs.Stripe,
+    customCheckoutSdk: stripeJs.StripeCustomCheckout
+  ) => {
+    setContext((ctx) => {
+      if (ctx.stripe && ctx.customCheckoutSdk) {
+        return ctx;
+      }
+
+      return {stripe, customCheckoutSdk};
+    });
+  };
+
+  React.useEffect(() => {
+    let isMounted = true;
+
+    if (parsed.tag === 'async' && !ctx.stripe) {
+      parsed.stripePromise.then((stripe) => {
+        if (stripe && isMounted && !initCustomCheckoutCalled) {
+          // Only update context if the component is still mounted
+          // and stripe is not null. We allow stripe to be null to make
+          // handling SSR easier.
+          setInitCustomCheckoutCalled(true);
+          stripe.initCustomCheckout(options).then((customCheckoutSdk) => {
+            if (customCheckoutSdk) {
```

Comment:
```suggestion
            if (customCheckoutSdk && isMounted) {
```

Reply:
After investigation, I don't we can/should check `isMounted` here.

After stripePromise is resolved, we sync set `initCustomCheckoutCalled` here https://github.com/stripe/react-stripe-js/blob/26bb4e222db70e47026f2fb03f99c72888374013/src/components/CustomCheckout.tsx#L129

This trigger the cleanup function before next render because the dependencies
https://github.com/stripe/react-stripe-js/blob/26bb4e222db70e47026f2fb03f99c72888374013/src/components/CustomCheckout.tsx#L152-L159

So if we check isMounted here, it's false when `initCustomCheckout` is resolved and it will NOT call safeSetContext https://github.com/stripe/react-stripe-js/blob/26bb4e222db70e47026f2fb03f99c72888374013/src/components/CustomCheckout.tsx#L132

Reply:
hm, the issue though is that you're potentially setting context values after the component has been unmounted. IIRC this throws a warning/error in the console?

Reply:
nvm they removed that warning at some point: https://github.com/reactwg/react-18/discussions/82

Reply:
> The cleanup function runs not only during unmount, but before every re-render with changed dependencies. 
https://react.dev/reference/react/useEffect#my-cleanup-logic-runs-even-though-my-component-didnt-unmount

The `isMounted` doesn't seem to be a good signal to tell whether the component is still mounted. I guess we could use `initCustomCheckoutCalled` state to signal it. The state is gone when unmount and doesn't tie to the changed dependencies. But I am not sure when it will be gone when unmount. 🤔